File: frontend/src/lib/audio.ts
/**
 * Fetch a fresh Deezer preview URL for the given track ID.
 * Deezer preview URLs expire; this fetches a new one on demand.
 */
export async function fetchFreshPreviewUrl(deezerTrackId: number): Promise<string> {
  const API_BASE = import.meta.env.VITE_API_URL || '/api'
  const res = await fetch(`${API_BASE}/audio/preview/${deezerTrackId}`)
  if (!res.ok) {
    throw new Error('Preview not available for this track')
  }
  const data = await res.json()
  const url = data?.preview_url
  if (!url || typeof url !== 'string') {
    throw new Error('Invalid preview URL response')
  }
  return url
}

File: backend/app/api/audio.py
"""Audio preview endpoints for just-in-time Deezer preview URL fetching."""

from fastapi import APIRouter, HTTPException

from app.services.deezer_service import fetch_fresh_preview_url

router = APIRouter(prefix="/audio", tags=["audio"])


@router.get("/preview/{deezer_track_id}")
def get_fresh_preview_url(deezer_track_id: int):
    """
    Fetch a fresh Deezer preview URL for the given track ID.
    Use this when playing audio to avoid expired cached URLs.
    """
    url = fetch_fresh_preview_url(deezer_track_id)
    if url is None:
        raise HTTPException(
            status_code=404,
            detail="Preview not available for this track",
        )
    return {"preview_url": url}
